File: bmodle/src/main/java/com/hong_world/bmodle/modle/TasksRepository.java
package com.hong_world.bmodle.modle;

import android.support.annotation.NonNull;

import com.hong_world.common.bean.FeedArticleListData;

import io.reactivex.Observable;

/**
 * Date: 2018/8/13. 10:48
 * Author: hong_world
 * Description:
 * Version:
 */
public class TasksRepository extends BaseTasksDataSource {
    private static TasksRepository INSTANCE = null;
    private final TasksDataSource mTasksRemoteDataSource;

    private final TasksDataSource mTasksLocalDataSource;

    /**
     * @param tasksRemoteDataSource
     * @param tasksLocalDataSource
     */
    private TasksRepository(@NonNull TasksDataSource tasksRemoteDataSource,
                            @NonNull TasksDataSource tasksLocalDataSource) {
        mTasksRemoteDataSource = tasksRemoteDataSource;
        mTasksLocalDataSource = tasksLocalDataSource;
    }

    public static TasksRepository getInstance(TasksDataSource tasksRemoteDataSource,
                                              TasksDataSource tasksLocalDataSource) {
        if (INSTANCE == null) {
            INSTANCE = new TasksRepository(tasksRemoteDataSource, tasksLocalDataSource);
        }
        return INSTANCE;
    }

    @Override
    public Observable<FeedArticleListData> getFeedArticleList(int num) {
        return mTasksRemoteDataSource.getFeedArticleList(num);
    }
}

File: kotlin-module/src/main/java/com/hong_world/kotlin_module/modle/TasksRepository.kt
package com.hong_world.kotlin_module.modle

import com.hong_world.kotlin_module.bean.FeedArticleListData
import io.reactivex.Observable

/**
 * Date: 2018/9/25. 16:28
 * Author: hong_world
 * Description:
 * Version:
 */
class TasksRepository : BaseTasksDataSource {

    private var mTasksRemoteDataSource: TasksDataSource? = null

    private var mTasksLocalDataSource: TasksDataSource? = null

    override fun getFeedArticleList(num: Int): Observable<FeedArticleListData>? = mTasksRemoteDataSource?.getFeedArticleList(num)

    private constructor(mTasksRemoteDataSource: TasksDataSource,
                        mTasksLocalDataSource: TasksDataSource) {
        this.mTasksLocalDataSource = mTasksLocalDataSource
        this.mTasksRemoteDataSource = mTasksRemoteDataSource

    }

    companion object {
        @JvmStatic
        fun getInstance(tasksRemoteDataSource: TasksDataSource,
                        tasksLocalDataSource: TasksDataSource): TasksRepository = Holder.getHolder(tasksRemoteDataSource, tasksLocalDataSource)
    }

    private object Holder {
        var tasksRepository: TasksRepository? = null
        fun getHolder(tasksRemoteDataSource: TasksDataSource,
                      tasksLocalDataSource: TasksDataSource): TasksRepository {
            if (tasksRepository == null) tasksRepository = TasksRepository(tasksRemoteDataSource, tasksLocalDataSource)
            return tasksRepository as TasksRepository
        }
    }
}
